### PR TITLE
feat: Add fine grained cache and topic level permissions to tokens

### PR DESCRIPTION
**Pull request by krispraws**
### Summary
- This PR adds the ability to scope tokens to certain topics or caches per the scenarios in this [doc](https://www.notion.so/momentohq/Authorization-Scenarios-f1cea5c1caee4523b3ebe85687c65a55).
- See related PR for proto changes: https://github.com/momentohq/client-protos/pull/179
- **There are breaking changes in `packages/core/src/auth/tokens/token-scope.ts`**
    - `Permission`, `CachePermission` and `TopicPermission` have been changed from `class` to `interface`
    - `None` has been removed from the `CacheRole` and `TopicRole` enums
    - `TopicRole.ReadWrite` has been renamed to `TopicRole.PublishSubscribe`
- New example for scoped tokens has been added in `examples/nodejs/access-control/access-control.ts`
However it has not been added to dependabot and workflows/build.yml because the CI jobs for testing examples do not build the sdk so we get errors like below. The new example will be added to workflows after the SDK is released.
```
/Users/runner/work/client-sdk-javascript/client-sdk-javascript/examples/nodejs/access-control/access-control.ts
Error:    18:8    error    Unable to resolve path to module '@gomomento/sdk'                import/no-unresolved
Error:    18:8    error    "@gomomento/sdk" is not found                                    node/no-missing-import
```

### Testing
- Added unit tests and integration tests

### CODE REVIEW

**Comment by malandis** [comment]
Looks good. Will approve once build failures are fixed.

File: packages/core/src/auth/tokens/token-scope.ts
Comment (krispraws) [inline packages/core/src/auth/tokens/token-scope.ts:7]: Specifically looking for approvals on names and shapes in this file today.

Comment (malandis) [reply packages/core/src/auth/tokens/token-scope.ts:7]: My apologies. I thought this was the client-protos PR!

File: packages/client-sdk-nodejs/src/internal/internal-auth-client.ts
Comment (cprice404) [inline packages/client-sdk-nodejs/src/internal/internal-auth-client.ts:302]: do we want to do this validation in the client? i almost think this would be better to just do on the server so that the errors are consistent across the different SDKs and we don't have to repeat the code so much?  just a thought.

File: packages/core/src/auth/tokens/token-scope.ts
Comment (cprice404) [inline packages/core/src/auth/tokens/token-scope.ts:7]: I don't feel strongly about this but if you liked `Any` better, you could call this `AnyCache`.

File: packages/core/src/auth/tokens/token-scope.ts
Comment (cprice404) [inline packages/core/src/auth/tokens/token-scope.ts:14]: 👍 to how you are doing the typing here where you are unioning the things together.

As for the name "Resource" here, I don't hate it, might have a slight pref for something like "Selector" or "Descriptor", but I don't love those either, and would be fine being outvoted

File: packages/core/src/auth/tokens/token-scope.ts
Comment (cprice404) [inline packages/core/src/auth/tokens/token-scope.ts:22]: nit, I think you could remove the `?` from the argument type and just put a default value directly in the signature?

File: packages/core/src/auth/tokens/token-scope.ts
Comment (cprice404) [inline packages/core/src/auth/tokens/token-scope.ts:51]: when you start having multiple optional positional args in JS it is common to create an interface to represent them. e.g.:

```
export interface TopicPermissionOpts {
   cache?: CacheResource,
   topic?: TopicResource
}
```

Then users can call it like this:

```
new TopicPermission(role, {topic: ...});
```

Otherwise it will feel weird that you are allowing `cache` to be optional but basically forcing them to pass `undefined` explicitly if they don't want to set it.

File: packages/core/src/auth/tokens/token-scope.ts
Comment (cprice404) [reply packages/core/src/auth/tokens/token-scope.ts:22]: if you end up making an Options interface for the Topic one as per my comment below, you will want to do it here too for consistency. In which case you won't be able to set a default value in the function signature and you can ignore my comment about that above :)

**Review by cprice404 — COMMENTED** [review]
handful of small things, food for thought, definitely open to discussion

File: packages/client-sdk-nodejs/src/internal/internal-auth-client.ts
Comment (krispraws) [reply packages/client-sdk-nodejs/src/internal/internal-auth-client.ts:302]: I was going to add the validation in both client and server, but I can remove it from the client if we have been following that pattern for other apis.

File: packages/core/src/auth/tokens/token-scope.ts
Comment (krispraws) [reply packages/core/src/auth/tokens/token-scope.ts:7]: Allen pointed out Any has special meaning in JS. I am fine with All - this way I don't have to create two messages `AnyCache` and `AnyTopic`

File: packages/core/src/auth/tokens/token-scope.ts
Comment (krispraws) [reply packages/core/src/auth/tokens/token-scope.ts:14]: I renamed Resource to Selector in the client-protos and am working on doing that here as well!

File: packages/core/src/auth/tokens/token-scope.ts
Comment (krispraws) [reply packages/core/src/auth/tokens/token-scope.ts:22]: Is one preferred over the other w.r.t. default values vs interface? I don't have strong opinions on Javascript so I can follow whatever the standards are.

Comment (cprice404) [reply packages/core/src/auth/tokens/token-scope.ts:22]: i would go with options interface, it is definitely the most common pattern when you have multiple optional arguments like you do in the topic one.

File: packages/core/src/auth/tokens/token-scope.ts
Comment (cprice404) [reply packages/core/src/auth/tokens/token-scope.ts:7]: that's fine with me but there will probably be some language where `All` is a keyword :P

we can just give it a different name in that SDK though if/when he hit that :)

File: packages/client-sdk-nodejs/src/internal/internal-auth-client.ts
Comment (cprice404) [reply packages/client-sdk-nodejs/src/internal/internal-auth-client.ts:302]: i wouldn't say it's a pattern, more of a case-by-case thing. but the more complex the validation is, the more concerned i am about the implementation drifting in one SDK vs the next vs the server. So I honestly might err towards just server-side on this one. but it's not a blocker for me

Comment (krispraws) [reply packages/client-sdk-nodejs/src/internal/internal-auth-client.ts:302]: Makes sense. Validation of this complexity is better off in the server.

**Comment by cprice404** [comment]
about to review this again but just curious what happens if this gets merged before the server-side changes?

File: packages/client-sdk-nodejs/test/unit/integration/auth-client.test.ts
Comment (cprice404) [inline packages/client-sdk-nodejs/test/unit/integration/auth-client.test.ts:2]: I apologize for this comment, I realize it's not something that was introduced in this PR and that I must have just missed in an earlier PR, my bad.

Two observations about this file:

1. The path says `unit/integration`. it should be one or the other. It looks like this is just a unit test? if so can we move it up a directory?
2. It's pretty unusual for us to be writing tests that directly reference the generated GRPC classes. Is there a way that we could get the same test coverage that are after here using only our public API classes? Such a test would be far more valuable in my eyes.

I won't block merging the PR on those points but I'd really like to address #2 in a fast follow if possible, if we don't do it in this PR. I don't want people to start copy/pasting the pattern of writing tests that reference the generated classes unless it's truly necessary :)

But please tell me if I'm wrong and these tests have high value and cannot be written using only public API classes.

File: packages/core/src/auth/tokens/token-scope.ts
Comment (cprice404) [inline packages/core/src/auth/tokens/token-scope.ts:14]: what would you think about supporting `| string` here and then having the code implicitly convert that to a `CacheName`?

File: packages/core/src/auth/tokens/token-scope.ts
Comment (cprice404) [inline packages/core/src/auth/tokens/token-scope.ts:45]: same comment here about `| string`

File: packages/core/src/auth/tokens/token-scope.ts
Comment (cprice404) [inline packages/core/src/auth/tokens/token-scope.ts:80]: not that I mind this being explicit but technically you shouldn't need to do this, right? Isn't it the default?

File: packages/core/src/auth/tokens/token-scope.ts
Comment (cprice404) [inline packages/core/src/auth/tokens/token-scope.ts:84]: can we add some test coverage of this stuff in the common-integration-tests code? Besides the value of the test coverage it would be helpful in terms of seeing some examples of what user code that consumes this will look like.

Comment (krispraws) [reply packages/core/src/auth/tokens/token-scope.ts:84]: Absolutely - I am working on those, and will add them to this PR as another commit

File: packages/core/src/auth/tokens/token-scope.ts
Comment (krispraws) [reply packages/core/src/auth/tokens/token-scope.ts:80]: yes, it is the default so not needed technically.

File: packages/client-sdk-nodejs/test/unit/integration/auth-client.test.ts
Comment (krispraws) [reply packages/client-sdk-nodejs/test/unit/integration/auth-client.test.ts:2]: 1. I agree - it confused me too when I saw the tests. I will move it up.
2. These are essentially unit tests for testing the conversion from SDK types to the client proto types so they have to use the generated GRPC classes. The integration tests will be the high value tests since those will implicitly test the conversions as well. If there was an api that would describe an existing token, I could have written a test that created a token and described it back and verified the permissions were the same.

**Comment by krispraws** [comment]
> about to review this again but just curious what happens if this gets merged before the server-side changes?



> about to review this again but just curious what happens if this gets merged before the server-side changes?

Things would break if anyone tried to generate a AllDataReadWrite token because it will now try to create the token with the new permission fields. But I don't plan to merge this before server side changes are deployed.

File: packages/core/src/auth/tokens/token-scope.ts
Comment (krispraws) [reply packages/core/src/auth/tokens/token-scope.ts:14]: I can do that. Curious why that is better? Is it so that you can do something like CacheSelector('foo')?

Comment (cprice404) [reply packages/core/src/auth/tokens/token-scope.ts:14]: I'm kind of just guessing because I don't think we have a place in the PR where you show examples of what user code will look like, but I think it will allow something like:

`new CachePermission(ReadOnly, 'my-awesome-cache')` instead of `new CachePermission(ReadOnly, new CacheName('my-awesome-cache')`.

and actually now that I think about it I think maybe CacheName should be an interface instead of a class. Maybe we can zoom about this for a second tomorrow morning?

File: packages/core/src/auth/tokens/token-scope.ts
Comment (cprice404) [reply packages/core/src/auth/tokens/token-scope.ts:84]: excellent, thanks

**Comment by cprice404** [comment]
> > about to review this again but just curious what happens if this gets merged before the server-side changes?
> 
> > about to review this again but just curious what happens if this gets merged before the server-side changes?
> 
> Things would break if anyone tried to generate a AllDataReadWrite token because it will now try to create the token with the new permission fields. But I don't plan to merge this before server side changes are deployed.

ok cool, in that case probably safest to convert this to a "draft" PR until then.

File: packages/client-sdk-nodejs/test/unit/integration/auth-client.test.ts
Comment (cprice404) [reply packages/client-sdk-nodejs/test/unit/integration/auth-client.test.ts:2]: okay i will defer to your judgment on #2.

File: packages/core/src/auth/tokens/token-scope.ts
Comment (krispraws) [reply packages/core/src/auth/tokens/token-scope.ts:14]: Sure, will ping you tomorrow morning

Comment (krispraws) [reply packages/core/src/auth/tokens/token-scope.ts:14]: Here are a couple of integration tests that show user code: https://github.com/momentohq/client-sdk-javascript/pull/598/files#diff-ae2e3efaec41bcc9a06f77ab73967c608830d8c711127a11f30253ca462a3fe3
I was waiting to test these with the server before I write more.

**Comment by cprice404** [comment]
@krispraws based on our call the other day and the thing you showed me where you were able to pass in an object literal to the function, I think you and I are on the same page, but please feel free to re-request review whenever you are ready for another pass.

File: packages/client-sdk-nodejs/test/unit/integration/auth-client.test.ts
Comment (krispraws) [reply packages/client-sdk-nodejs/test/unit/integration/auth-client.test.ts:2]: Moved the unit test out of the `unit/integration` folder.

File: packages/core/src/auth/tokens/token-scope.ts
Comment (krispraws) [reply packages/core/src/auth/tokens/token-scope.ts]: Removed the explicit arguments for All and also the redundant unit test I had added, since the unit test for `AllDataReadWrite` will also test we are using the right defaults.

File: packages/core/src/auth/tokens/token-scope.ts
Comment (krispraws) [reply packages/core/src/auth/tokens/token-scope.ts]: Integration tests added.

**Comment by krispraws** [comment]
> > > about to review this again but just curious what happens if this gets merged before the server-side changes?
> > 
> > 
> > > about to review this again but just curious what happens if this gets merged before the server-side changes?
> > 
> > 
> > Things would break if anyone tried to generate a AllDataReadWrite token because it will now try to create the token with the new permission fields. But I don't plan to merge this before server side changes are deployed.
> 
> ok cool, in that case probably safest to convert this to a "draft" PR until then.

I was actually wrong about things breaking. Control plane would just have ignored the new fields in the proto, so nothing would break.

File: packages/core/src/auth/tokens/token-scope.ts
Comment (krispraws) [reply packages/core/src/auth/tokens/token-scope.ts]: Changed CacheName and TopicName to interface instead of class. Also added string as a member of the CacheSelector and TopicSelector unions.

File: packages/core/src/auth/tokens/token-scope.ts
Comment (krispraws) [reply packages/core/src/auth/tokens/token-scope.ts]: done

File: packages/common-integration-tests/src/auth-client.ts
Comment (cprice404) [inline packages/common-integration-tests/src/auth-client.ts:401]: idk if these 2 publish calls really do anything for the test?

File: packages/common-integration-tests/src/auth-client.ts
Comment (cprice404) [inline packages/common-integration-tests/src/auth-client.ts:471]: nit: reading these tests I'm having a little trouble mapping "foo" and "bar" to FGA_CACHE_* without scrolling back and forth

File: packages/common-integration-tests/src/auth-client.ts
Comment (cprice404) [inline packages/common-integration-tests/src/auth-client.ts:495]: same nit re: foo/bar in these comment strings

File: packages/common-integration-tests/src/auth-client.ts
Comment (cprice404) [inline packages/common-integration-tests/src/auth-client.ts:554]: would love to see one more test that did ReadOnly on topics

File: packages/core/src/auth/tokens/token-scope.ts
Comment (cprice404) [inline packages/core/src/auth/tokens/token-scope.ts:7]: now that I see that the class definition for `All` is empty, I think you can just do `export const` instead of `export class`, and then you won't need the `new All` stuff. You will have to choose a type for it that can be disambiguated from `String` though. could be an enum or an integer. or you can still use a class and just make the class private, and have `All` be a constant instance of the class.

**Review by cprice404 — COMMENTED** [review]
looking really good! can't wait to play with it.

Just a few nits/suggestions/questions

File: packages/common-integration-tests/src/auth-client.ts
Comment (krispraws) [reply packages/common-integration-tests/src/auth-client.ts:401]: They are just creating existing topics on a cache. One of the tests tries to publish to both existing and new topics and ensures both fail when you don't have topic permissions.

Comment (cprice404) [reply packages/common-integration-tests/src/auth-client.ts:401]: topics aren't a resource that actually exists or has state on the server though, so I don't think these lines are needed.

File: packages/client-sdk-nodejs/test/unit/auth-client.test.ts
Comment (krispraws) [reply packages/client-sdk-nodejs/test/unit/auth-client.test.ts]: Done in https://github.com/momentohq/client-sdk-javascript/pull/598/commits/f2aad35500cbec9f6c42ca858e70abafffacbf58

File: packages/common-integration-tests/src/auth-client.ts
Comment (krispraws) [reply packages/common-integration-tests/src/auth-client.ts]: Removed in https://github.com/momentohq/client-sdk-javascript/pull/598/commits/f2aad35500cbec9f6c42ca858e70abafffacbf58

File: packages/common-integration-tests/src/auth-client.ts
Comment (krispraws) [reply packages/common-integration-tests/src/auth-client.ts]: Fixed in https://github.com/momentohq/client-sdk-javascript/pull/598/commits/f2aad35500cbec9f6c42ca858e70abafffacbf58

File: packages/common-integration-tests/src/auth-client.ts
Comment (krispraws) [reply packages/common-integration-tests/src/auth-client.ts:596]: Added in https://github.com/momentohq/client-sdk-javascript/pull/598/commits/f2aad35500cbec9f6c42ca858e70abafffacbf58

File: packages/core/src/auth/tokens/token-scope.ts
Comment (krispraws) [reply packages/core/src/auth/tokens/token-scope.ts]: Done in https://github.com/momentohq/client-sdk-javascript/pull/598/commits/f2aad35500cbec9f6c42ca858e70abafffacbf58

**Review by cprice404 — DISMISSED** [review]
I'm good with this PR at this point; my only note:

I don't want to do a release of the JS SDK until we have spent some time bug-bashing this with an eye on the API, just in case we come up with any other usability changes.

and it would be great if we can add an example in the examples dir, and possibly add web sdk support, prior to release.

we don't have any urgent things that I know of that need to get out in the JS SDK so I think it's okay to merge and block release for a few days or whatever, but we just need to make sure everyone's aware of that plan once we merge this.

**Review by cprice404 — APPROVED** [review]
same caveat I mentioned in my previous approval about not wanting to do a release of the SDK with this just yet, but I'm fine with merging

**Comment by krispraws** [comment]
> I'm good with this PR at this point; my only note:
> 
> I don't want to do a release of the JS SDK until we have spent some time bug-bashing this with an eye on the API, just in case we come up with any other usability changes.
> 
> and it would be great if we can add an example in the examples dir, and possibly add web sdk support, prior to release.
> 
> we don't have any urgent things that I know of that need to get out in the JS SDK so I think it's okay to merge and block release for a few days or whatever, but we just need to make sure everyone's aware of that plan once we merge this.

Added an example in https://github.com/momentohq/client-sdk-javascript/pull/598/commits/8480434030e88fee8d8fb67964104614b471131d 
It is very basic and works on my machine.
Re: web-sdk support, the work is likely minimal but I would need to scope it. @danielamiao - do we want to add web sdk support right now?

**Comment by danielamiao** [comment]
@krispraws @cprice404 let's not make web sdk support a release blocker, but a fast follow post release.

File: examples/nodejs/access-control/package.json
Comment (cprice404) [inline examples/nodejs/access-control/package.json:28]: this is the right thing to do for this PR but we should come back and change it to a release version as soon as we have one.

We should also add this new example dir to CI, in the .github actions for PRs and for dependabot.

Those do not need to block this PR but we should do them when we can.

**Review by cprice404 — DISMISSED** [review]

File: examples/nodejs/access-control/package.json
Comment (krispraws) [reply examples/nodejs/access-control/package.json:28]: Do we have a runbook for everything that needs to be done for releasing new SDKs? It would be really useful.

Comment (cprice404) [reply examples/nodejs/access-control/package.json:28]: this isn't really about new SDKs or releases, it's about new examples. we should add something in CONTRIBUTING.md.

Comment (cprice404) [reply examples/nodejs/access-control/package.json:28]: https://github.com/momentohq/client-sdk-javascript/pull/620

File: examples/nodejs/access-control/README.md
Comment (poppoerika) [inline examples/nodejs/access-control/README.md:18]: I thought we want to encourage users to use our console to generate a token?

Comment (krispraws) [reply examples/nodejs/access-control/README.md:18]: Good catch, I copied this from the examples/nodejs/cache/README.md

**Comment by krispraws** [comment]
> > I'm good with this PR at this point; my only note:
> > I don't want to do a release of the JS SDK until we have spent some time bug-bashing this with an eye on the API, just in case we come up with any other usability changes.
> > and it would be great if we can add an example in the examples dir, and possibly add web sdk support, prior to release.
> > we don't have any urgent things that I know of that need to get out in the JS SDK so I think it's okay to merge and block release for a few days or whatever, but we just need to make sure everyone's aware of that plan once we merge this.
> 
> Added an example in [8480434](https://github.com/momentohq/client-sdk-javascript/commit/8480434030e88fee8d8fb67964104614b471131d) It is very basic and works on my machine. Re: web-sdk support, the work is likely minimal but I would need to scope it. @danielamiao - do we want to add web sdk support right now?

web-sdk changes are now part of this PR - thanks to @rishtigupta !

File: examples/nodejs/access-control/README.md
Comment (NoSQLKnowHow) [inline examples/nodejs/access-control/README.md:18]:
```suggestion
- A Momento auth token is required, you can generate one using the [Momento CLI](https://github.com/momentohq/momento-cli)
```

No upper case for "auth token" as it is not a product name.

File: packages/core/src/auth/tokens/token-scope.ts
Comment (krispraws) [inline packages/core/src/auth/tokens/token-scope.ts:43]: Do we want to change the enum values as well for consistency?

File: examples/nodejs/access-control/README.md
Comment (krispraws) [reply examples/nodejs/access-control/README.md]: fixed in https://github.com/momentohq/client-sdk-javascript/pull/598/commits/d583b64ee0db14557dc380df1c86785c67dbd326

Comment (krispraws) [reply examples/nodejs/access-control/README.md]: fixed in https://github.com/momentohq/client-sdk-javascript/pull/598/commits/d583b64ee0db14557dc380df1c86785c67dbd326

File: packages/core/src/auth/tokens/token-scope.ts
Comment (malandis) [reply packages/core/src/auth/tokens/token-scope.ts:40]: +1 from me --a developer may print a enum value be surprised by the inconsistency.

**Review by malandis — COMMENTED** [review]
Looks good! Just one suggestion and I agree we should change the enum values to be consistent.

File: packages/client-sdk-nodejs/test/unit/cache-client.test.ts
Comment (krispraws) [reply packages/client-sdk-nodejs/test/unit/cache-client.test.ts]: This change snuck in because of a merge from main followed by revert. I have merged the changes from main again.
Fixed in https://github.com/momentohq/client-sdk-javascript/pull/598/commits/bd1add565f4a5c62a79606fdc11edc43904bb02e

File: packages/core/src/auth/tokens/token-scope.ts
Comment (krispraws) [reply packages/core/src/auth/tokens/token-scope.ts]: Fixed in https://github.com/momentohq/client-sdk-javascript/pull/598/commits/649148592f9a49d01715fa7258ff2bee2c8e741d

**Review by malandis — APPROVED** [review]
🎉

**Review by rishtigupta — APPROVED** [review]

**Review by malandis — DISMISSED** [review]

File: packages/client-sdk-web/src/index.ts
Comment (bruuuuuuuce) [inline packages/client-sdk-web/src/index.ts:98]: do we also need to export `TokenScopes` here too?

**Review by bruuuuuuuce — COMMENTED** [review]
Looks great! One comment about exports for the web client, but otherwise LGTM

File: packages/client-sdk-web/src/index.ts
Comment (krispraws) [reply packages/client-sdk-web/src/index.ts:99]: Good catch! fixed in https://github.com/momentohq/client-sdk-javascript/pull/598/commits/37fef37d3c8a3a47ee92595a9d3070de4bbf8d1c

**Review by bruuuuuuuce — APPROVED** [review]
:shipit:

**Pull request merged**
